components/eam/src/physics/cam/cldera_dynamic_tracers.F90

```diff
@@ -0,0 +1,371 @@
+!===============================================================================
+! CLDERA dynamic tracers
+! provides interfaces for registering and adding histories for dynamic tracer constituents
+!
+! Potential vorticity tracer described in:
+! Whitehead, J.P. and Jablonowski, C. (2015) 
+! Potential vorticity: Measuring consistency between GCM dynamical cores and tracer advection 
+! schemes: Potential Vorticity: Measuring Consistency, Quarterly Journal of the Royal 
+! Meteorological Society, doi:10.1002/qj.2389
+!
+!===============================================================================
+
+module cldera_dynamic_tracers
+
+  use shr_kind_mod, only: r8 => shr_kind_r8
+  use spmd_utils,   only: masterproc
+  use ppgrid,       only: pcols, pver
+  use constituents, only: pcnst, cnst_add, cnst_name, cnst_longname
+  use cam_logfile,  only: iulog
+  use ref_pres,     only: pref_mid_norm
+
+  implicit none
+  private
+  save
+
+  ! Public interfaces
+  public :: cldera_dynamic_tracers_register        ! register constituents
+  public :: cldera_dynamic_tracers_implements_cnst ! true if constituent is implemented by this package
+  public :: cldera_dynamic_tracers_init            ! initialize history fields, datasets
+  public :: cldera_dynamic_tracers_init_cnst       ! initialize constituent field
+  public :: cldera_dynamic_tracers_timestep_tend   ! compute timestep tendencies
+  public :: cldera_dynamic_tracers_readnl          ! read namelist options
+  public :: cldera_dynamic_tracers_is_enabled      ! getter function for checking if package is activated
+  public :: cldera_dynamic_tracers_pv_idx          ! getter function for PV tracer global constituent index
+  public :: cldera_dynamic_tracers_pt_idx          ! getter function for PT tracer global constituent index
+  public :: cldera_dynamic_tracers_is_pt           ! return true if named tracer is PT tracer
+  public :: cldera_dynamic_tracers_is_pv           ! return true if names tracer is PV tracer
+  public :: cldera_dynamic_tracers_pv_offset       
+
+  real(r8), parameter :: cldera_dynamic_tracers_pv_offset = 1_r8
+
+  ! ----- Private module data
+  integer, parameter :: ncnst=2  ! number of constituents implemented by this module
+
+  ! constituent names
+  character(len=8), parameter :: c_names(ncnst) = (/'PV_TRCR_   ', 'PT_TRCR    '/)
```

Comment:
Why does `PV_TRCR_` have a trailing underscore by `PT_TRCR` does not?

Reply:
My understanding is that the EAM infrastructure defaultly provides the ability to output constituent quantities once you have registered them. As PV_TRCR_ is offset from the correct PV tracer value by a constant factor, the trailing underscore in "PV_TRCR_" denotes this distinction from "PV_TRCR", which is the derived quantity. This was done because I was usure if whether writing a derived quantity into the constitutent output field would be overwritten by EAM.

Reply:
Ok, I went and reviewed how the constituents are registered, and "cam_outfld=.false." is being passed to the PV initialization but not the PT initialization. Therefore I think we could replace all references to "PV_TRCR_" with "PV_TRCR" and everything would be fine.

Reply:
Okay I’m a little confused then; is PV_TRCR_ the offset tracer? I thought the offset was removed before being output.

Reply:
Hey @OkayHughes , I guess I still don't understand this. Are you saying that we don't need this extra field at all anymore? So can we completely remove references to `pv_out_name`, `pv_idx`, and `PV_TRCR_`, replacing them with those that are equivalent to the handling of PT? 

@brhillman I believe what is happening is that `PV_TRCR_` is the offset tracer, and it is advected independently of `PV_TRCR`, which is just a placeholder field. At output, the values of `PV_TRCR_` has their offset subtracted off; this recovers the correct PV values, and the result is written to `PV_TRCR`. 

I think that all Owen is saying is that he didn't know if EAM would allow him to overwrite the values of an existing tracer quantity (or at least make them negative), so instead of advecting and then shifting `PV_TRCR`, he decided to advect an auxiliary quantity PV_TRCR_, but ultimately write out `PV_TRCR`, as desired.

In his comments here, he suggests that this was not a necessary thing to do, and we can get rid of that auxiliary field and simplify all of this, I think.

@OkayHughes if you can confirm that this is correct, I'll make the necessary changes.

Reply:
This is what I was trying to convey. I won’t be able to test the changed code myself, but my best understanding is that changing the code so the offset field (now called “PV_TRCR_”) to be called “PV_TRCR” will result in identical behavior but remove this ambiguity

Reply:
Okay, I will remove `PV_TRCR_`, and also references to `pv_idx` in favor of the q-field

Reply:
Scratch that, I actually wasn't totally sure if the removal of the PV offset (in `cldera_dynamic_tracers_timestep_tend`) happened before or after advection. Rather than having to take the time to do runs and test this, I opted to just leave it as is. I think it is actually appropriately legible to have the offset PV function as a separate auxiliary field, assuming this is described in the comments.

I added some inline comments to `cldera_dynamic_tracers.F90`, near the declaration of `c_names`, which describes the approach and the difference between the fields. There is also a new comment which describes the initial application of the offset in `dyn_comp.F90`, as mentioned in another resolved issue. With the presence of these comments, I think the code is improved.